Fix 593840: MetadataReferenceCache implementation does IO on the UI thread

DavidKarlas opened the pull request:
Also moved from Timer checking to FileSystemWatcher to observe referenced .dlls changes, added support for removing references and implemented IMetadataService.

Commit:
Fix 593840: MetadataReferenceCache implementation does IO on the UI thread
Also moved from Timer checking to FileSystemWatcher to observe referenced .dlls changes, added support for removing references and implemented IMetadataService.

mrward commented on main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.TypeSystem/MetadataReferenceCache.cs:
```diff
@@ -212,6 +238,51 @@ protected override string GetDocumentationForSymbol (string documentationMemberI
+				public FileChangeTracker (string filePath)
+				{
+					FilePath = filePath;
+					watcher = new FileSystemWatcher (Path.GetDirectoryName (filePath), Path.GetFileName (filePath));
```
Unfortunately right now the native file system watcher does not seem to generate change events correctly. The Mono/.NET file watcher works as expected. Although in this case things may just work since it has event handlers for all possible events.

mrward replied:
Adding to @Therzok suggestions:
c) Move the FileChangeTracker class to Core. Then it can use the native file system watcher and the IDE can use the FileChangeTracker class from Core.
d) Wait for Mono 5.14.

mkrueger reviewed (dismissed):
The fsw should work most of the time - anyways we need that 100% working in mono. 

Some things in the VS.NET world rely on that. So if it do esn't work in some cases we should fix it in mono.

slluis commented:
If the watcher problem is fixed in Mono 5.14, let's set it as required dependency. Dismissing approval until this is fixed.

Therzok commented:
@slluis that means we can't merge it until the next release after 15.8. As mono master has not even been branched yet.

slluis commented:
@Therzok Mono 5.14 is branched, it's the 2018-04 branch, which goes to stable in 15.8.

Therzok commented:
:+1: from me then

Therzok commented:
I'm taking this over while David is on holidays. I've already written support for smarter caching by reusing roslyn code, which uses mmap'd files rather than pure IO (this allows for metadata references to be constructed from pointer offsets rather than reading the full file).

Therzok commented:
So, I did some research around corefx file watchers in mono, and it seems like 5.14 is returning empty filesystemwatcher handle

Therzok reviewed (dismissed):
Requesting changes until I land vs roslyn import and figure out why Mono FSW is broken

Therzok committed:
[IDE] Import Roslyn VS code for MetadataReference usage

This commit changes a few more things in the MetadataReferenceCache:

* Together with caching the MetadataReference, it also caches the Metadata
into a temporary storage stream. This, in turn, has the bonus of not having
to load the full assembly, but having direct pointer access to the assembly's
internals

* Ties the MetadataReferenceCache to the workspace and mirrors how VS uses it

* Fixes a few locking issues around ProjectData

* MetadataReferences no longer need to manually be removed, they're tied
to ProjectData and ProjectData removal means reference removal

* ProjectData is now properly removed, so project reloads will not cause
infinite growing list of ProjectData

* Unit tests for basic functionality

* Uses a MetadataReference Resolver that resolves the full path to the
referenced assembly in case relative paths are used

* Unhacks users of MetadataReferenceCache

Therzok commented:
See https://github.com/mono/mono/issues/8712

Therzok committed:
Add FileChangeTracker in core, unit tests now work

Therzok committed:
Deterministically fix unit tests

Therzok commented:
Going green soon

Therzok commented:
Found a bug.

Therzok committed:
Fix race on metadata references updating

Therzok commented:
Need to rewrite the file watching code to handle most common paths via normal watching (workspace, runtime assemblies) and create watchers for out of tree assembly folders.

Therzok committed:
[Roslyn] Port MetadataReferenceManager over to FileWatcherService

Therzok commented:
Reminder: Keep track of https://github.com/dotnet/roslyn/pull/28162

Therzok committed:
[Roslyn] Track laziness changes around metadata references.

Therzok commented:
Reminder for myself tomorrow, to look into providing perf-sensitive APIs around PathTreeNode without probing for paths, and registrations (i.e. metadata references) avoid tree lookups (which is basically `O(logn)`)

Therzok commented:
Track lazy fsw subscription: https://github.com/dotnet/roslyn/pull/28475/files

Therzok commented:
https://github.com/dotnet/roslyn/pull/28490/files

Therzok committed:
[Forms] Fix tests

Therzok commented:
This PR is ready for testing now!

The main fixes is:
a) We should not be refreshing assemblies when focusing out/in the IDE, causing minor chokes while typing
b) We should test that when referencing an assembly by dll (not project) in a multi project setup:

```
project A
class MainClass
{
    public static void Main ()
    {
        MyOtherClass.Do();
    }
}
```

```
project B
public class MyOtherClass
{
    public static void Do() {}
}
```

where A references B by assembly (not project). Build solution, change B to be:
```
project B
public class MyOtherClass
{
    //public static void Do() {}
}
```

Build just B, see that A has build errors in the IDE

mkrueger commented on main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.TypeSystem/MonoDevelopWorkspace.cs:
```diff
@@ -1493,7 +1560,9 @@ protected override void ApplyMetadataReferenceRemoved (ProjectId projectId, Meta
-			if (mdProject == null || path == null)
+			if (path == null)
+				return;
+			if (mdProject == null)
```
I find the old version easier to read :)

Therzok replied:
I'm not sure how this got in. I'll fix this.

Therzok committed:
Style fix

mkrueger commented on main/tests/Ide.Tests/MonoDevelop.Ide.TypeSystem/MonoDevelopMetadataReferenceManager.cs:
```diff
@@ -0,0 +1,261 @@
+			int size;
+			using (var copyStream = SerializableBytes.CreateWritableStream ()) {
+				// open a file and let it go as soon as possible
+				using (var fileStream = FileUtilities.OpenRead (moduleFileKey.FullPath)) {
```
Can you put that in try…catch file operations may file for some reason - as a general pattern - just to have some info in the log.

Therzok replied:
This is caught in `MonoDevelopMetadataReference.Snapshot.GetMetadataImpl`. If we catch here, the error will not bubble when it should.

mrward commented on main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.TypeSystem/MonoDevelopWorkspace.cs:
```diff
@@ -556,7 +610,21 @@ public DocumentId GetDocumentId (string name)
+
+				lock (workspace.updatingProjectDataLock) {
+					MonoDevelop.Projects.FileWatcherService.WatchDirectories (this, null).Ignore ();
+					lock (metadataReferences) {
```
Is this needed? It does not seem that the ProjectData adds itself to the file watcher.

Therzok committed:
[Ide] Remove superfluous unwatch

Therzok commented:
Rebased and all fixed!

mrward reviewed (approved):
Tested out b) described in comment - 
https://github.com/mono/monodevelop/pull/4683#issuecomment-406575296 - and all is good with a direct assembly reference and rebuilding the other project.

Therzok commented:
@monojenkins backport release-7.6

monojenkins commented:
@Therzok backporting to release-7.6 failed, the patch results in conflicts:

```
Applying: Fix 593840: MetadataReferenceCache implementation does IO on the UI thread Also moved from Timer checking to FileSystemWatcher to observe referenced .dlls changes, added support for removing references and implemented IMetadataService.
Using index info to reconstruct a base tree...
M	main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.TypeSystem/MonoDevelopWorkspace.cs
M	main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.csproj
Falling back to patching base and 3-way merge...
Auto-merging main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.csproj
CONFLICT (content): Merge conflict in main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.csproj
Auto-merging main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.TypeSystem/MonoDevelopWorkspace.cs
error: Failed to merge in the changes.
Patch failed at 0001 Fix 593840: MetadataReferenceCache implementation does IO on the UI thread Also moved from Timer checking to FileSystemWatcher to observe referenced .dlls changes, added support for removing references and implemented IMetadataService.
```

Please backport manually!

sgmunn commented:
@Therzok don't forget to backport to 7.7 as well
